internal/controller/testrun_controller.go

```diff
@@ -0,0 +1,405 @@
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/clock"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	etosv1alpha1 "github.com/eiffel-community/etos/api/v1alpha1"
+)
+
+const (
+	typeActive      = "Active"
+	typeSuiteRunner = "SuiteRunner"
+	typeEnvironment = "Environment"
+)
+
+// TestRunReconciler reconciles a TestRun object
+type TestRunReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Clock  clock.WithTicker
+}
+
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=testruns,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=testruns/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=testruns/finalizers,verbs=update
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=environments,verbs=get;watch;create;delete
+// +kubebuilder:rbac:groups=etos.eiffel-community.github.io,resources=environments/status,verbs=get
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups=batch,resources=jobs/status,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/reconcile
+func (r *TestRunReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.V(2).Info("Get testrun", "namespace", req.Namespace, "name", req.Name)
+	testrun := &etosv1alpha1.TestRun{}
+	if err := r.Get(ctx, req.NamespacedName, testrun); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("Testrun not found, exiting", "namespace", req.Namespace, "name", req.Name)
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "Testrun not found", "namespace", req.Namespace, "name", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	requeue, err := r.reconcile(ctx, testrun, req)
+	if err != nil {
+		if errors.IsConflict(err) {
+			return ctrl.Result{Requeue: true}, nil
```

Comment:
This assumes that any conflict error shall be solved by requeue. Is it correct?

Reply:
It is correct.
Because of how controller operate, the `status` field may be updated by other controllers. If the `status` field has been updated by another we will get a conflict. That is what this check is catching. A requeue is the proper way to handle it.